ssl/statem/statem_clnt.c

```diff
@@ -2454,7 +2450,6 @@ MSG_PROCESS_RETURN tls_process_new_session_ticket(SSL *s, PACKET *pkt)
     if (ticklen == 0)
         return MSG_PROCESS_CONTINUE_READING;
 
-    /* TODO(TLS1.3): Is this a suitable test for TLS1.3? */
```

Comment:
I don't believe this check is sufficient for TLS 1.3. It will cause thread-safety bugs in some applications. Consider an application using `SSL_get_session` written before TLS 1.3. This may happen:

```
Thread A:
  ssl = SSL_new();
  SSL_do_handshake(ssl);
  StoreInSessionCache(SSL_get_session(ssl)) // Stores a ticket-less session
  SSL_read(ssl) // discovers a post-handshake NewSessionTicket

Thread B:
  ssl2 = SSL_new();
  SSL_set_session(ssl2, LoadFromSessionCache());
  SSL_do_handshake(ssl2) // in progress
```

A's `StoreInSessionCache` passes the session to B's `LoadFromSessionCache`. `SSL_read` now fills the ticket into that `SSL_SESSION` while B is busy using it.

The necessary invariant is: any `SSL_SESSION` that a caller may plausibly put into the session cache must immutable. In BoringSSL, our TLS 1.3 NewSessionTicket code always duplicates the session for this reason.